examples/step-62/step-62.cc

```diff
@@ -0,0 +1,1467 @@
+        Point<dim>(data.get_attribute<double>("probe_pos_x"),
+                   data.get_attribute<double>("probe_pos_y") -
+                     data.get_attribute<double>("probe_width_y") / 2))
+    , probe_stop_point(
+        Point<dim>(data.get_attribute<double>("probe_pos_x"),
+                   data.get_attribute<double>("probe_pos_y") +
+                     data.get_attribute<double>("probe_width_y") / 2))
+    , right_hand_side(data)
+    , pml(data)
+    , rho(data)
+  {}
+
+
+
+  // @sect4{`QuadratureCache` class}
+
+  // We need to reserve enough space for the mass and stiffness matrices and the
+  // right hand side vector.
+  template <int dim>
+  QuadratureCache<dim>::QuadratureCache(unsigned int dofs_per_cell)
+    : dofs_per_cell(dofs_per_cell)
+    , mass_coefficient(dofs_per_cell, dofs_per_cell)
+    , stiffness_coefficient(dofs_per_cell, dofs_per_cell)
+    , right_hand_side(dofs_per_cell)
+  {}
+
+
+
+  // @sect3{Implementation of the `ElasticWave` class}
+
+  // @sect4{Constructors and destructors}
+
+  // This is very similar to the constructor of step-40. In addition we create
+  // the HDF5 datasets `frequency_dataset`, `position_dataset` and
+  // `displacement`. Note the use of the `template` for the creation of the HDF5
+  // datasets. It is a C++ requirement to use the `template` keyword in order to
+  // treat `create_dataset` as a dependent template name.
+  template <int dim>
+  ElasticWave<dim>::ElasticWave(const Parameters<dim> &parameters)
+    : parameters(parameters)
+    , mpi_communicator(MPI_COMM_WORLD)
+    , triangulation(mpi_communicator,
+                    typename Triangulation<dim>::MeshSmoothing(
+                      Triangulation<dim>::smoothing_on_refinement |
+                      Triangulation<dim>::smoothing_on_coarsening))
+    , quadrature_formula(2)
+    , n_q_points(quadrature_formula.size())
+    , dof_handler(triangulation)
+    , fe(FE_Q<dim>(1), dim)
+    , frequency(parameters.nb_frequency_points)
+    , position(parameters.nb_probe_points, dim)
+    , frequency_dataset(parameters.data.template create_dataset<double>(
+        "frequency",
+        std::vector<hsize_t>{parameters.nb_frequency_points}))
+    , position_dataset(parameters.data.template create_dataset<double>(
+        "position",
+        std::vector<hsize_t>{parameters.nb_probe_points, dim}))
+    , displacement(
+        parameters.data.template create_dataset<std::complex<double>>(
+          "displacement",
+          std::vector<hsize_t>{parameters.nb_probe_points,
+                               parameters.nb_frequency_points}))
+    , pcout(std::cout,
+            (Utilities::MPI::this_mpi_process(mpi_communicator) == 0))
+    , computing_timer(mpi_communicator,
+                      pcout,
+                      TimerOutput::summary,
+                      TimerOutput::wall_times)
+  {}
+
+
+
+  template <int dim>
+  ElasticWave<dim>::~ElasticWave()
+  {
+    dof_handler.clear();
+  }
```

Comment:
Do you actually need this? Usually, if you just declare the finite element object *before* the DoFHandler, then you don't need the call to `clear()`, and consequently don't need a destructor at all.

Reply:
I copied this code from step-40. I can add a PR to modify step-40 accordingly :)

I think that it would be good to document in DoFHandler.clear() that there is no need to call this function in the destructor if the finite element object has been declared before the dof_handler. I've not see it in the documentation.

Reply:
Sure, if you feel like modifying step-40 in a separate PR, we'd take that as well :-)

Thinking about it, it doesn't actually depend on the order of the objects any more. That used to be the case, but it's no longer the case. This should simply *always* work without the call to `clear()`.